From 22d8f4ce41625021d4fc173dff0df8b87b982d7d Mon Sep 17 00:00:00 2001
From: Lukas Czerner <lczerner@redhat.com>
Date: Thu, 23 Oct 2014 09:54:35 -0400
Subject: [fs] ext4: translate fallocate mode bits to strings

Message-id: <1414058125-4183-122-git-send-email-lczerner@redhat.com>
Patchwork-id: 98233
O-Subject: [RHEL 7.1 PATCH v2 121/171] ext4: translate fallocate mode bits to strings
Bugzilla: 1150139
RH-Acked-by: Eric Sandeen <sandeen@redhat.com>

BZ 1150139
https://bugzilla.redhat.com/show_bug.cgi?id=1150139

BREW 8142710
https://brewweb.devel.redhat.com/taskinfo?taskID=8142710

Upstream commit a633f5a319cf4116d977e25fea2830dce23a8e74

 Signed-off-by: Lukas Czerner <lczerner@redhat.com>
 Signed-off-by: "Theodore Ts'o" <tytso@mit.edu>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/ext4/ext4.h b/fs/ext4/ext4.h
index 5ddb38d..bd127b7 100644
--- a/fs/ext4/ext4.h
+++ b/fs/ext4/ext4.h
@@ -31,6 +31,7 @@
 #include <linux/percpu_counter.h>
 #include <linux/ratelimit.h>
 #include <crypto/hash.h>
+#include <linux/falloc.h>
 #ifdef __KERNEL__
 #include <linux/compat.h>
 #endif
diff --git a/fs/ext4/extents.c b/fs/ext4/extents.c
index e0f3b30..052b27e 100644
--- a/fs/ext4/extents.c
+++ b/fs/ext4/extents.c
@@ -37,7 +37,6 @@
 #include <linux/quotaops.h>
 #include <linux/string.h>
 #include <linux/slab.h>
-#include <linux/falloc.h>
 #include <asm/uaccess.h>
 #include <linux/fiemap.h>
 #include "ext4_jbd2.h"
diff --git a/include/trace/events/ext4.h b/include/trace/events/ext4.h
index deb12ac..f7a280e 100644
--- a/include/trace/events/ext4.h
+++ b/include/trace/events/ext4.h
@@ -68,6 +68,11 @@ struct extent_status;
  { EXTENT_STATUS_DELAYED, "D" },   \
  { EXTENT_STATUS_HOLE,  "H" })
 
+#define show_falloc_mode(mode) __print_flags(mode, "|",  \
+ { FALLOC_FL_KEEP_SIZE,  "KEEP_SIZE"},  \
+ { FALLOC_FL_PUNCH_HOLE,  "PUNCH_HOLE"},  \
+ { FALLOC_FL_NO_HIDE_STALE, "NO_HIDE_STALE"})
+
 
 TRACE_EVENT(ext4_free_inode,
  TP_PROTO(struct inode *inode),
@@ -1349,10 +1354,10 @@ TRACE_EVENT(ext4_fallocate_enter,
   __entry->mode = mode;
  ),
 
- TP_printk("dev %d,%d ino %lu pos %lld len %lld mode %d",
+ TP_printk("dev %d,%d ino %lu pos %lld len %lld mode %s",
     MAJOR(__entry->dev), MINOR(__entry->dev),
     (unsigned long) __entry->ino, __entry->pos,
-    __entry->len, __entry->mode)
+    __entry->len, show_falloc_mode(__entry->mode))
 );
 
 TRACE_EVENT(ext4_fallocate_exit,
-- 
1.7.1